### ISSUE TITLE
Merge clipa into main

### ISSUE THREAD

**Issue by rom1504** [mlfoundations/open_clip#558]
https://github.com/UCSC-VLAA/CLIPA/tree/master/clipa_torch code is there

Would be nice to identify the changes and PR them here

**Comment by zw615** [comment]
Hi! This is Zeyu, one of the authors of CLIPA. I have listed the changes of CLIPA of follows. And if it all looks good, I can start a pull request if suitable. 

##################

**Inference**

As for the CLIPA-v2 trained weights, they are converted from our Jax codebase. This implementation is heavily based on the BigVision codebase, which therefore has several differences compared to OpenCLIP. To make our models runable with OpenCLIP codebase, the following changes need to be made:


1. _"gelu_approximate": "tanh"_.  Use a GELU activation function with approximation. Should be implemented with 2 lines of code (https://github.com/UCSC-VLAA/CLIPA/blob/3f91974161f66e66b6c446b159933bbeebb6d4b6/clipa_torch/open_clip/model.py#L129)


2. _"ln_pre": false_. No LN layer after patch dropout. Should be implemented with 1 line of code (https://github.com/UCSC-VLAA/CLIPA/blob/3f91974161f66e66b6c446b159933bbeebb6d4b6/clipa_torch/open_clip/transformer.py#L390)


3. _"pool_style": "big_vision_gap"_. The pooling style of ViT in BigVision is a little bit different, where cls token is excluded. Should be implemented with 5 lines of code (https://github.com/UCSC-VLAA/CLIPA/blob/3f91974161f66e66b6c446b159933bbeebb6d4b6/clipa_torch/open_clip/transformer.py#L521)


4. _"vocab_size": 32000_. The vocabulary size of the bert tokenizer used in BigVision is 32000. 


5. _"bert_tokenizer": true. "vocab_path": "gs://vit_models/lit/bert/uncased_L-12_H-768_A-12/vocab.txt"_. Use BERT tokenizer. Should be implemented with 40 lines of code (https://github.com/UCSC-VLAA/CLIPA/blob/3f91974161f66e66b6c446b159933bbeebb6d4b6/clipa_torch/open_clip/factory.py#L86C28-L86C28)


6. _"pool_style": "big_vision_last"_.  The pooling style of the text transformer in BigVision is a little bit different. Should be implemented with 5 lines of code. (https://github.com/UCSC-VLAA/CLIPA/blob/3f91974161f66e66b6c446b159933bbeebb6d4b6/clipa_torch/open_clip/transformer.py#L668)


7. _"attention_mask": false_. BigVision does not use an attention mask. Should be implemented with 5 lines of code. (https://github.com/UCSC-VLAA/CLIPA/blob/3f91974161f66e66b6c446b159933bbeebb6d4b6/clipa_torch/open_clip/transformer.py#L589)


There are two other differences about the data augmentation our jax implementation used:

_--square-resize-only_. Resize both sides to the same size. Should be implemented with 5 lines of code (https://github.com/UCSC-VLAA/CLIPA/blob/3f91974161f66e66b6c446b159933bbeebb6d4b6/clipa_torch/open_clip/transform.py#L191)

_--interpolation 'bilinear'_. Use bilinear interpolation. Should be implemented with 1 line of code. (https://github.com/UCSC-VLAA/CLIPA/blob/3f91974161f66e66b6c446b159933bbeebb6d4b6/clipa_torch/open_clip/transform.py#L195)

#########################

In general, the modifications should not be very hard to integrate, and they do not have a significant impact on the final performance in our observations. Let me know if you have any other questions.

**Comment by rom1504** [comment]
the key point here is to implement all of this without impacting the default functioning of the repo
eg by default everything should work exactly the same as now, all models should continue working

that means figuring out what flags to use and how to organize this

**Comment by rom1504** [comment]
what about training of clipa models? that's the most interesting part

**Comment by zw615** [comment]
Yes. Of course you implement CLIPA without impacting the default functioning of the OpenCLIP repo! That should be quite easy, as the core idea of CLIPA is a multi-stage training. 

At stage 1, use reduced lengths input. For image input, use the `--force-image-size` option or modify the "image_size" attribute (e.g. 84 for H/14) in the corresponding json file in `model_configs`. For text input, modify the "context_length" attribute (e.g. 8) in the corresponding model json file, and use the syntax masking. It prioritizes retaining nouns, followed by adjectives, and then other words. You may find the masking code in [tokenizer.py](https://github.com/UCSC-VLAA/CLIPA/blob/3f91974161f66e66b6c446b159933bbeebb6d4b6/clipa_torch/open_clip/tokenizer.py#L281), which consists of about 60 lines of code.

At stage 2 (or 3, if there is any), use full size input. Again, change the value for the "image_size" and  "context_length" attributes.

Another thing I want to mention is that we found using global average pooling, rather than cls token, works better for larger models during CLIPA training. 

In general, you can configure most changes in the model config json file and the runnning script. Check our [model config files](https://github.com/UCSC-VLAA/CLIPA/blob/master/clipa_torch/open_clip/model_configs/ViT-L-16-CL8-GAP.json) for details. The GELU, LN, tokenizer, pool_style, attention_mask, and data augmentation changes only matter if the trained CLIPA-v2 weights (converted from jax) are used.

**Comment by zw615** [comment]
BTW, I can initiate two pull requests, one for training, and one for inference. There should be some overlap but not much.

**Comment by rom1504** [comment]
Sounds good!

If it's possible to split in small independent PRs that are easy to merge
that's even better. (We have some "too big to merge" PRs open in this repo
and it's best if we can avoid this situation)

**Comment by mehdidc** [comment]
Great @zw615! please go ahead, I could then proceed and try to reproduce some of  the experiments

**Comment by zw615** [comment]
Hi @mehdidc @rom1504, the pr for clipa training is added. Some minor details (e.g. sincos2d pos_embed, pooling_style) which we found to be insignificant in our preliminary exps are not included. There could be some variations to our results but not much. Let me know if you have any questions.

Also, I am working on the pr for clipa inference with the converted clipa weights. Should be finished within a day or two.

**Comment by zw615** [comment]
clipa inference pr also pushed
